### ISSUE TITLE
Possible to expose functions in b3RobotSimulatorClientAPI? 

### ISSUE THREAD

**Issue by evandarcy** [kripken/ammo.js#256]
@kripken  I have already exposed some functions I need on my fork of ammo.js, it wasn't so hard as these functions were listed in the bullet API documentation so I could just copy and paste.

However, I came across a function that is not referenced in the Bullet API documentation called _setJointMotorControl()_ in this program on line 35: https://github.com/bulletphysics/bullet3/blob/master/examples/RobotSimulator/MinitaurSetup.cpp

         for (int i = 0; i < numJoints; i++)
	{
		b3RobotSimulatorJointMotorArgs controlArgs(CONTROL_MODE_VELOCITY);
		controlArgs.m_maxTorqueValue = 0;
		sim->setJointMotorControl(m_data->m_quadrupedUniqueId, i, controlArgs);
	}

It seems like this function is declared in this file on line 542: https://github.com/bulletphysics/bullet3/blob/cdd56e46411527772711da5357c856a90ad9ea67/examples/SharedMemory/b3RobotSimulatorClientAPI_NoDirect.h

Is it even possible to add functions referenced in the RobotSimulator API to Ammo.js? If anybody can help me on this one, it would be awesome because I'm fairly stumped!

**Comment by willeastcott** [comment]
I believe ammo.js is built off of Bullet 2.8.3 so you can only expose Bullet API from that version (not Bullet 3 as you have above).

**Comment by evandarcy** [comment]
Hey Will! Do you know if there is a specific reason it hasn't been updated to use the latest Bullet (Version 2.88)?

I found this fork of Ammo that is supposedly using Bullet 3 (Version 2.8.5): https://github.com/WhitestormJS/AmmoNext

There are .cpp files in the AmmoNext repo above that reference the function _setJointMotorControl()_  that I want. In theory is it possible to expose that function in the ammo.idl file? I would understand how to do it if it was a normal function with prefix "bt" but I'm just not sure how I would go about doing it for a function with prefix "b3" that is not referenced in API documentation.

**Comment by willeastcott** [comment]
I don't know why @kripken hasn't updated to the latest 2.x version. I don't see a reason not to other than, hey, it just takes time. But updating to Bullet 3 could, I guess, cause some behavioral difference in the physics which might break some peoples' applications?

**Comment by kripken** [comment]
Updating to 2.x is just time, yeah - I'm too busy myself, sorry. Would be great to see it done though!

**Comment by evandarcy** [comment]
@kripken , I'd be happy to give it a go. How much work is actually involved, surely Emscripten does most of the work right?

I've been trying to find a tutorial online on how to convert a library using Emscripten, I found one here: https://developers.google.com/web/updates/2018/03/emscripting-a-c-library , I did try to apply the same steps to bullet to no success.

Perhaps you could point me in the right direction or give me a procedure to follow? 

Thanks!

**Comment by evandarcy** [comment]
Hey @kripken , I got Ammo to build with bullet3. I am trying to expose a function but am getting this error message when I try to use the exposed function constructor.

<img width="1318" alt="Screen Shot 2019-05-31 at 22 40 55" src="https://user-images.githubusercontent.com/31148326/58736368-72565f00-83f5-11e9-83c8-8897fb611c5b.png">

Have you every seen an error like this before? The constructor is b3RobotSimulatorClientAPI_NoDirect(), I have no idea what _ZN34b3RobotSimulatorClientAPI_NoDirectC1Ev is...

**Comment by evandarcy** [comment]
I also get this warning in stage 6 of python make.py 
`warning: undefined symbol: _ZN34b3RobotSimulatorClientAPI_NoDirectC1Ev
`

**Comment by kripken** [comment]
Basically, the procedure would be to replace the current bullet source files in this repo with the newer ones, then build and see how things go. If APIs change, then we'd need to also update the `idl` file, the examples, and the tests.

**Comment by evandarcy** [comment]
So I have successfully managed to build using bullet3 and was able to expose a function from bullet3 that returns an API version and run in browser successfully.

Now I am trying to expose this "connect" function
<img width="713" alt="Screen Shot 2019-06-06 at 20 08 13" src="https://user-images.githubusercontent.com/31148326/59059432-dfa33d80-8896-11e9-88f9-df04d1f1b9ed.png">

By placing this in the .idl file
<img width="581" alt="Screen Shot 2019-06-06 at 20 05 36" src="https://user-images.githubusercontent.com/31148326/59059465-f2b60d80-8896-11e9-8c86-c65e17e89418.png">

And I am trying to call it in a html file like so 
<img width="528" alt="Screen Shot 2019-06-06 at 20 09 48" src="https://user-images.githubusercontent.com/31148326/59059514-0f524580-8897-11e9-8329-fab7aa1019f4.png">

But I am getting the following error in browser
<img width="992" alt="Screen Shot 2019-06-06 at 20 10 20" src="https://user-images.githubusercontent.com/31148326/59059546-22651580-8897-11e9-9565-1b3bf6973661.png">

Is this an error you recognise @kripken ?

**Comment by kripken** [comment]
Not sure what that could be. The stack trace in a debug build might be more useful though.

**Comment by evandarcy** [comment]
Rather confused..how do I go about doing a debug build?

**Comment by kripken** [comment]
Probably simplest to edit `make.py` and add `-g` after `-O3`, so `args = '-O3 -g --llvm-lto 1` etc.

**Comment by evandarcy** [comment]
Okay, so I did that. Here's something interesting I found:

This is the new error I am getting: 

<img width="993" alt="Screen Shot 2019-06-08 at 13 04 34" src="https://user-images.githubusercontent.com/31148326/59147034-ff994500-89ed-11e9-9a40-d34176980662.png">

When I click into it, I can see the following (Notice mention of FS):

<img width="634" alt="Screen Shot 2019-06-08 at 13 05 33" src="https://user-images.githubusercontent.com/31148326/59147038-2192c780-89ee-11e9-8d5b-3719078df057.png">

When building project using python make.py, I get the following "undefined symbols" errors. Notice ($FS):
<img width="1437" alt="Screen Shot 2019-06-08 at 13 02 55" src="https://user-images.githubusercontent.com/31148326/59147031-e7292a80-89ed-11e9-9a60-fa4f768401e8.png">

I have had to fix a lot of these "undefined symbol" errors by tracking down which library the function mentioned in the symbol belongs to and including the correct library file (.o) file for that function in the make.py file. It was a very repetitive process and I am not sure it was the correct thing to do but I was able to eliminate each "undefined symbol" error one by one by including correct library.

Perhaps the error I am getting in console has to do with $FS being excluded somehow in a similar sense as other undefined symbols?

**Comment by evandarcy** [comment]
Update: I set SYSCALLS_REQUIRE_FILESYSTEM equal to 0 in emscripten settings.js and rebuilt, i no longer get an error for $FS. When I try to run same example in browser I get this error instead:

<img width="1316" alt="Screen Shot 2019-06-08 at 13 26 19" src="https://user-images.githubusercontent.com/31148326/59147237-0d9c9500-89f1-11e9-8533-23f3e823b5ab.png">

**Comment by evandarcy** [comment]
@kripken I am consistently running into errors to do with filesystem. Do I need to add a --preload-file somewhere in the make.py file to include files I want to use?
<img width="1317" alt="Screen Shot 2019-06-09 at 19 31 52" src="https://user-images.githubusercontent.com/31148326/59163072-9ac81280-8af3-11e9-96ee-69ab4f720256.png">

**Comment by kripken** [comment]
I tested latest emscripten incoming now, and the ammo tests pass for me. We did have some FS related bugs in the past, which emscripten version are you on? (Or did you add code in JS to use files somehow, like that error suggests?)

**Comment by evandarcy** [comment]
Yes I was trying to load a .urdf file, I think I've figured out the FS stuff, no longer getting errors there. Now the problem lies in figuring out how to load a .urdf file in webGL.

**Comment by beiller** [comment]
Wow @evandarcy you need to give us a Pull Request with your bullet3!
